src/Core/src/Handlers/View/AbstractViewHandler.Android.cs

```diff
@@ -46,27 +46,44 @@ public void SetFrame(Rectangle frame)
 
 		public virtual Size GetDesiredSize(double widthConstraint, double heightConstraint)
 		{
-			if (TypedNativeView == null)
+			if (TypedNativeView == null || VirtualView == null || Context == null)
 			{
 				return Size.Zero;
 			}
 
-			if (Context == null)
-			{
-				return new Size(widthConstraint, heightConstraint);
-			}
-
-			var deviceWidthConstraint = Context.ToPixels(widthConstraint);
-			var deviceHeightConstraint = Context.ToPixels(heightConstraint);
-
-			var widthSpec = MeasureSpecMode.AtMost.MakeMeasureSpec((int)deviceWidthConstraint);
-			var heightSpec = MeasureSpecMode.AtMost.MakeMeasureSpec((int)deviceHeightConstraint);
+			// Create a spec to handle the native measure
+			var widthSpec = CreateMeasureSpec(widthConstraint, VirtualView.Width);
+			var heightSpec = CreateMeasureSpec(heightConstraint, VirtualView.Height);
 
 			TypedNativeView.Measure(widthSpec, heightSpec);
 
+			// Convert back to xplat sizes for the return value
 			return Context.FromPixels(TypedNativeView.MeasuredWidth, TypedNativeView.MeasuredHeight);
 		}
 
+		int CreateMeasureSpec(double constraint, double explicitSize)
+		{
+			var mode = MeasureSpecMode.AtMost;
+
+			if (explicitSize > -1)
+			{
+				// We have a set value (i.e., a Width or Height)
+				mode = MeasureSpecMode.Exactly;
+				constraint = explicitSize;
+			}
+			else if (double.IsInfinity(constraint))
```

Comment:
Do we want to do anything for NaN?

Reply:
Nope. NaN isn't a valid value for measuring; the rest of these are.